### PR TITLE
fix: sort and quote keys with Yarn's rules when writing yarn.lock

### PR DESCRIPTION
**Pull request by shalvah**
This PR adjusts the logic used for sorting keys when writing `yarn.lock` to be more in line with Yarn's own rules, avoiding unnecessary diffs and confusion.

## References
Addresses some of the issues raised in #5126, by mostly adapting the code Yarn itself uses. It doesn't solve everything (I didn't address the registry URLs issue, and sorting differs slightly when there are `~` or `^`), but it aligns this more closely with what Yarn produces. If the npm team is interested, I can work on PRs to try to address the others.

### CODE REVIEW

**Comment by cachius** [comment]
@npm/cli-team This would be really cool to have in v9 (and 8.19.3)!
@shalvah I'd sure be interested in more fixes.

File: workspaces/arborist/lib/yarn-lock.js [hunk 1]
```diff
@@ -33,10 +32,45 @@ const consistentResolve = require('./consistent-resolve.js')
...
+const priorityThenLocaleCompare = (a, b) => {
+  if (yarnEntryPriorities[a] || yarnEntryPriorities[b]) {
+    return (yarnEntryPriorities[a] || 100) > (yarnEntryPriorities[b] || 100) ? 1 : -1
+  }
+
+  return localeCompare(a, b)
+}
+
+const shouldQuoteString = str => {
+  return str.indexOf('true') === 0 ||
```
Comment (wraithgar) [inline workspaces/arborist/lib/yarn-lock.js:55]: is there a reason this isn't a test of equality with the string `true` or `false`?

File: workspaces/arborist/lib/yarn-lock.js [hunk 2]
```diff
@@ -33,10 +32,45 @@ const consistentResolve = require('./consistent-resolve.js')
...
+      str.indexOf('false') === 0 ||
+      /[:\s\n\\",[\]]/g.test(str) ||
+      /^[0-9]/g.test(str) ||
+      !/^[a-zA-Z]/g.test(str)
```
Comment (wraithgar) [inline workspaces/arborist/lib/yarn-lock.js:59]: If it starts with 0-9 then it doesn't start w/ a-zA-Z so the 0-9 check isn't needed.

File: workspaces/arborist/lib/yarn-lock.js [hunk 3]
```diff
@@ -33,10 +32,45 @@ const consistentResolve = require('./consistent-resolve.js')
...
+}
+
+const quoteIfNeeded = val => {
+  if (typeof val === 'boolean' || typeof val === 'number' || shouldQuoteString(val)) {
```
Comment (wraithgar) [inline workspaces/arborist/lib/yarn-lock.js:63]: idk if we need two functions here. single use `shouldQuoteString` logic can be inlined w/ the rest of this if statement.

**Comment by wraithgar** [comment]
Tweaked and consolidated things in https://github.com/npm/cli/pull/5724 in the interest of getting this out w/ today's release.

**Comment by wraithgar** [comment]
> If the npm team is interested, I can work on PRs to try to address the others.

Yes please @shalvah anyone who is willing to own this is more than welcome to collaborate w/ us on keeping npm up to spec.

**Comment by shalvah** [comment]
Alright @wraithgar . Thanks for picking it up. Guess I'll close this PR then. Currently in the hospital with a broken arm so I couldn't make any changes. 😓Will hopefully work on the future improvements when I'm better.

**Comment by wraithgar** [comment]
Hope you get better soon. Looking forward to future PRs.

**Pull request closed without merging**